File: src/main/java/com/anod/appwatcher/tags/TagAppsManager.java
package com.anod.appwatcher.tags;

import android.content.Context;
import android.database.Cursor;
import android.support.v4.util.SimpleArrayMap;

import com.anod.appwatcher.content.DbContentProvider;
import com.anod.appwatcher.content.DbContentProviderClient;
import com.anod.appwatcher.model.Tag;
import com.anod.appwatcher.model.schema.AppTagsTable;
import com.anod.appwatcher.utils.CollectionsUtils;

import java.util.ArrayList;

class TagAppsManager {
    private SimpleArrayMap<String, Boolean> mApps = new SimpleArrayMap<>();
    private boolean mDefaultSelected;
    private final Tag mTag;
    private final Context mContext;

    TagAppsManager(Tag tag, Context context) {
        mTag = tag;
        mContext = context;
    }

    void selectAll(boolean select) {
        mApps.clear();
        mDefaultSelected = select;
    }

    void updateApp(String appId, boolean checked) {
        mApps.put(appId, checked);
    }

    boolean isSelected(String appId) {
        if (mApps.containsKey(appId)) {
            return mApps.get(appId);
        }
        return mDefaultSelected;
    }

    void initSelected(Cursor data) {
        if (data == null || data.getCount() == 0) {
            return;
        }
        data.moveToPosition(-1);
        while (data.moveToNext()) {
            String appId = data.getString(AppTagsTable.Projection.APPID);
            mApps.put(appId, true);
        }
        data.close();
    }

    boolean runImport() {
        ArrayList<String> appIds = new ArrayList<>();
        for (int i = 0; i <mApps.size(); i++)
        {
            if (mApps.valueAt(i)) {
                appIds.add(mApps.keyAt(i));
            }
        }

        DbContentProviderClient cr = new DbContentProviderClient(mContext);
        boolean result = cr.addAppsToTag(appIds, mTag);
        cr.close();

        mContext.getContentResolver().notifyChange(DbContentProvider.APPS_TAG_CONTENT_URI, null);
        return result;
    }
}

File: src/main/java/com/anod/appwatcher/tags/TagAppsManager.kt
package com.anod.appwatcher.tags

import android.content.Context
import android.database.Cursor
import android.support.v4.util.SimpleArrayMap
import com.anod.appwatcher.content.DbContentProvider
import com.anod.appwatcher.content.DbContentProviderClient
import com.anod.appwatcher.model.Tag
import com.anod.appwatcher.model.schema.AppTagsTable
import java.util.*

internal class TagAppsManager(private val mTag: Tag, private val mContext: Context) {
    private val mApps = SimpleArrayMap<String, Boolean>()
    private var mDefaultSelected: Boolean = false

    fun selectAll(select: Boolean) {
        mApps.clear()
        mDefaultSelected = select
    }

    fun updateApp(appId: String, checked: Boolean) {
        mApps.put(appId, checked)
    }

    fun isSelected(appId: String): Boolean {
        if (mApps.containsKey(appId)) {
            return mApps.get(appId)
        }
        return mDefaultSelected
    }

    fun initSelected(data: Cursor?) {
        if (data == null || data.count == 0) {
            return
        }
        data.moveToPosition(-1)
        while (data.moveToNext()) {
            val appId = data.getString(AppTagsTable.Projection.APPID)
            mApps.put(appId, true)
        }
        data.close()
    }

    fun runImport(): Boolean {
        val appIds = (0..mApps.size() - 1)
                .filter { mApps.valueAt(it) }
                .map { mApps.keyAt(it) }

        val cr = DbContentProviderClient(mContext)
        val result = cr.setAppsToTag(appIds, mTag.id)
        cr.close()

        mContext.contentResolver.notifyChange(DbContentProvider.APPS_TAG_CONTENT_URI, null)
        return result
    }
}